### PR TITLE
exporter: fix instance_id labeld value for rgw in exporter

### PR DESCRIPTION
**Pull request by avanthakkar**
Follows: https://github.com/ceph/ceph/pull/52191

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by avanthakkar** [comment]
jenkins test make check

**Comment by avanthakkar** [comment]
jenkins test make check

**Comment by avanthakkar** [comment]
jenkins test make check

File: src/exporter/DaemonMetricCollector.cc [hunk 1]
```diff
@@ -306,13 +306,15 @@ DaemonMetricCollector::get_labels_and_metric_name(std::string daemon_name,
     labels["instance_id"] = quote(tmp);
   }
   else if (daemon_name.find("rgw") != std::string::npos) {
-    // fetch intance_id for e.g. "okbvtv" from daemon_name=rgw.foo.ceph-node-00.okbvtv 
-    size_t pos = daemon_name.find_last_of(".");
```
Comment (pereman2) [inline src/exporter/DaemonMetricCollector.cc:310]: you can use [find_nth](https://www.boost.org/doc/libs/1_54_0/doc/html/boost/algorithm/find_nth.html) to get the third 4th element.

**Comment by avanthakkar** [comment]
jenkins test make check

File: src/exporter/DaemonMetricCollector.cc (hunk 1 above)
Comment (jmolmo) [reply src/exporter/DaemonMetricCollector.cc:310]: Good simplification. It seems a good idea :-).

**Review by jmolmo — CHANGES_REQUESTED** [review]

File: src/exporter/DaemonMetricCollector.cc [hunk 2]
```diff
@@ -306,13 +306,19 @@ DaemonMetricCollector::get_labels_and_metric_name(std::string daemon_name,
...
+    std::stringstream ss;
+    ss.str(daemon_name);
+    std::string item;
+    while (std::getline(ss, item, '.')) {
+        elems.push_back(item);
+    }
+    if (elems.size() >= 4) {
+      labels["instance_id"] = quote(elems[3]);
+    } else {
+      derr << "Error: Not enough elements in `elems` to parse instance_id from daemon_name=" << daemon_name << dendl;
```
Comment (jmolmo) [inline src/exporter/DaemonMetricCollector.cc:320]: Could it be useful to add the item when not located?
`
labels["instance_id"] = 'unknown'; `

Comment (avanthakkar) [reply src/exporter/DaemonMetricCollector.cc:320]: > Could it be useful to add the item when not located? `labels["instance_id"] = 'unknown';`

We shouldn't add any labels if it doesn;t exist in prometheus.

Comment (avanthakkar) [reply src/exporter/DaemonMetricCollector.cc:320]: And btw PR is still in draft

**Comment by avanthakkar** [comment]
jenkins test make check

**Review by jmolmo — APPROVED** [review]

**Comment by jmolmo** [comment]
jenkins test make check

**Comment by jmolmo** [comment]
jenkins test make check arm64

**Comment by avanthakkar** [comment]
jenkins test make check

**Comment by avanthakkar** [comment]
jenkins test api

**Comment by avanthakkar** [comment]
jenkins test dashboard cephadm

**Comment by avanthakkar** [comment]
jenkins test dashboard

**Comment by avanthakkar** [comment]
jenkins test dashboard

**Comment by avanthakkar** [comment]
jenkins retest this please

**Comment by avanthakkar** [comment]
jenkins test dashboard

**Pull request merged**